kamu-cli/src/services/workspace_service.rs

```diff
@@ -0,0 +1,165 @@
+            }
+        }
+    }
+
+    /// Returns the version that code expects to function correctly
+    pub fn latest_supported_version(&self) -> usize {
+        WorkspaceLayout::VERSION
+    }
+
+    /// Perform an upgrade of the workspace if necessary
+    pub fn upgrade(&self) -> Result<WorkspaceUpgradeResult, WorkspaceUpgradeError> {
+        let prev_version = self
+            .workspace_version()?
+            .expect("Upgrade called when not in workspace");
+
+        let mut current_version = prev_version;
+        let new_version = WorkspaceLayout::VERSION;
+
+        if prev_version == new_version {
+            return Ok(WorkspaceUpgradeResult {
+                prev_version,
+                new_version,
+            });
+        }
+
+        tracing::info!(%prev_version, %new_version, "Upgrading workspace");
+
+        while current_version != new_version {
+            tracing::info!(
+                "Upgrading from version {} to {}",
+                current_version,
+                new_version
+            );
+
+            match current_version {
+                0 => self.upgrade_0_to_1()?,
+                _ => unreachable!(),
+            }
+
+            current_version += 1;
+            std::fs::write(
+                &self.workspace_layout.version_path,
+                current_version.to_string(),
+            )
+            .int_err()?;
+        }
+
+        Ok(WorkspaceUpgradeResult {
+            prev_version,
+            new_version,
+        })
+    }
+
+    fn upgrade_0_to_1(&self) -> Result<(), InternalError> {
+        for entry in self.workspace_layout.datasets_dir.read_dir().int_err()? {
+            let dataset_dir = entry.int_err()?;
+            let cache_dir = dataset_dir.path().join("cache");
+            if cache_dir.is_dir() {
+                tracing::info!(?cache_dir, "Deleting old dataset ingest cache directory");
+                std::fs::remove_dir_all(cache_dir).int_err()?;
+            }
+        }
+        let _ = std::fs::create_dir(&self.workspace_layout.cache_dir);
+        Ok(())
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct WorkspaceUpgradeResult {
+    pub prev_version: usize,
+    pub new_version: usize,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum WorkspaceUpgradeError {
+    #[error(transparent)]
+    FutureVersion(#[from] WorkspaceFutureVersionError),
+    #[error(transparent)]
+    Internal(
+        #[from]
+        #[backtrace]
+        InternalError,
+    ),
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(thiserror::Error, Debug)]
+#[error("Workspace version {workspace_version} is newer than supported version {latest_supported_version} - upgrade to latest software version")]
+pub struct WorkspaceFutureVersionError {
+    pub workspace_version: usize,
```

Comment:
Maybe an enum instead of `usize` ?
The values could be still integers, but the named labels might be useful and more expressive.

Reply:
Done! Followed a similar pattern as you had for `MetadataBlockVersion`.